submariner/globalnet-enhancement2-0.md

```diff
@@ -0,0 +1,312 @@
+
+On the other hand, when the backend Pod tries to talk to the external application in remote
+Cluster, the external application will see that traffic is coming with a globalIP that is
+assigned either at the Cluster level, or namespace level, or the one requested by the user
+at Pod level. This asynchronous behavior will be analogous to the behavior seen with standard
+K8s applications when they talk to any Services. Applications desiring to have the same IPaddress
+for both ingress and egress traffic can create Headless Services in front of them and should
+avoid requesting explicit globalIP that match Headless Service backend Pods.
+
+The proposal has its own set of Pros and Cons.
+
+### Pros
+
+Since we will be annotating only select K8s objects with globalIP, it will help in improved scalability.
+Also, this translates to fewer iptable rules on the active Gateway node of the cluster which will
+improve the overall performance of the solution.
+
+### Cons
+
+A limitation with the proposed approach is that we will lose the unique identity for **every**
+Pod which is supported today. For users that intend to use Network Policies across Clusters or
+for debugging purposes, this can be mitigated by creating the necessary NamespaceGlobalEgressIP CRDs
+which create unique globalIPs at the Namespace/Pod level.
+
+### Caveats
+
+In the current implementation, when a client Pod is scheduled, until it's annotated with a globalIP,
+and the corresponding egress rules are programmed on the node, it was not able to communicate with
+remote Services. This was creating some issues in-terms of user-experience. In this proposal, as we
+have a globalIP at the cluster level, a Pod can communicate with remote Services right away which
+will be closer to the Vanilla Submariner experience.
+
+## Design Details
+
+The following CRD is proposed to support the new design.
+
+```Go
+type ClusterGlobalEgressIP struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec is the specification of desired behavior of ClusterGlobalEgressIP object.
+    Spec               ClusterEgressIPSpec `json:"spec"`
+
+    // Observed status of NamespaceGlobalEgressIP. Its a read-only field.
+    // +optional
+    Status             EgressIPStatus `json:"status,omitempty"`
+}
+
+type ClusterEgressIPSpec struct {
+    // The number of globalIP's requested. Globalnet Controller will allocate the requested
+    // number of contiguous GlobalIPs for this ClusterGlobalEgressIP object.
+    // If unspecified, NumGlobalIPs defaults to 1.
+    // +optional
+    NumGlobalIPs  int     'json:"numGlobalIPs",omitempty"`
+
+    // Selects the pods whose label match the definition. This is an optional field and
+    // in case it is not set, it results in all the Pods selected from all Namespaces.
+    // +optional
+    PodSelector metav1.LabelSelector `json:"podSelector,omitempty"`
+}
+
+type NamespaceGlobalEgressIP struct {
```

Comment:
```suggestion
type GlobalEgressIP struct {
```

I would suggest that we follow the convention of other models in K8s, Role vs ClusterRole, RoleBinding vs ClusterRoleBinding, etc.

In our case it would be GlobalEgressIP vs ClusterGlobalEgressIP

Also makes the name shorter.

Reply:
Ack. Will do, thanks for reviewing @mangelajo